### ISSUE TITLE
Parallel inserts on duplicate values may cause all shard placements to be FILE_INACTIVE

### ISSUE THREAD

**Issue by metdos** [citusdata/citus#401]
Assume that your replication factor is 2, and you are trying to insert the same value with **unique** constraint from two clients at the same time. This is what happens; Client1 inserts into Shard1, then Client2 tries to insert into Shard1 and fails. Client2 insert into Shard2 before Client1, and when Client1 tries to insert into Shard2 it fails. As a result, Client1 marks Shard2 as inactive, and Client2 marks Shard1 as inactive.

You can see ExecuteDistributedModify() for details.

Here is how to replicate the problem;
```sql
CREATE TABLE temp (
        my_time timestamp,
primary key (my_time));

SELECT master_create_distributed_table('temp', 'my_time', 'hash');

SELECT master_create_worker_shards('temp', 16, 2);
```

Run this bash script in parallel. In seconds, you should be able to see inactive shard palacements in pg_dist_shard_placement.

```bash
#!/bin/bash
for i in `seq 1 10000`;
do
    psql -d postgres -c "INSERT INTO temp (my_time) values ('$(date)')"
done
```

**Comment by marcocitus** [comment]
Related to #484

**Comment by ozgune** [comment]
@metdos -- is this issue also related to #221?

**Comment by marcocitus** [comment]
I think the permanent fix for this issue, which also considers connection failures, would be to take an exclusive lock before changing placement metadata. #221 would solve this issue for query failures.

**Comment by jasonmp85** [comment]
I'll verify this with some Docker images shortly.

**Comment by jasonmp85** [comment]
BTW I ran this on a docker-compose cluster using the `5.1.0` tag and the `nightly` tag from yesterday and I could reproduce the failure on `5.1.0` but on `nightly` no placements are marked inactive (what we expect!)

So looks like this bug is fixed.

**Comment by metdos** [comment]
I also verified on my box that this problem is solved.

**Comment by marcocitus** [comment]
Could we open another issue for other parallel insert failures?
